cordapp-upgrades/README.md

```diff
@@ -0,0 +1,158 @@
+<p align="center">
+  <img src="https://www.corda.net/wp-content/uploads/2016/11/fg005_corda_b.png" alt="Corda" width="500">
+</p>
+
+# The Upgrade CorDapp
+
+This directory contains a number of versions of the same CorDapp (based heavily on the Obligation sample CorDapp). It is
+designed to showcase ways in which a CorDapp might be upgraded such that the app remains backwards compatible between
+versions.
+
+The CorDapp is divided into contracts and workflows. These are versioned separately.
+
+# Current versions
+
+The CorDapp contract and workflow versions are described below.
+
+## Contracts
+
+### Version 1
+
+The initial contract version.
+
+### Version 2
+
+This second version adds a new field to the obligation state that allows the borrower to default on the obligation. A
+command is added to the contract to facilitate this. The new field has to be nullable in order to ensure backwards
+compatibility between nodes with the old contract version and nodes with the new contract version while a rolling
+upgrade is carried out.
+
+## Workflows
+
+### Version 1
+The initial version of the CorDapp. This is the Obligation CorDapp using the version of FinalityFlow that was present in
+Corda 3.
+
+Compiled against version 1 of the contracts CorDapp.
+
+### Version 2
+This version of the CorDapp upgrades to use the new version of the FinalityFlow present in Corda 4. To do this such that
+ the app remains compatible with Version 1, the following must be done:
+ - The flows using the new FinalityFlow must be versioned
+ - The flows must detect if the counterparty is using the old version, and revert to the old API accordingly
+   
+Compiled against version 1 of the contracts CorDapp.
+
+### Version 4
+This version adds the flows required to default on an Obligation.
+
+Compiled against version 2 of the contracts CorDapp.
+
+# Upgrade scenarios
+
+This section described the upgrade scenarios that can be explored using the workflows and contracts in this CorDapp.
+Note that all commands should be run from the root of the CorDapp directory structure (i.e. the location of this Readme).
+
+## Scenario 1: Finality Flow Upgrade
+
+This scenario covers upgrading an app to run on V4 by upgrading its usage of FinalityFlow to the new API defined in
```

Comment:
Add link to https://docs.corda.net/head/app-upgrade-notes.html#step-5-security-upgrade-your-use-of-finalityflow